File: src/main/java/com/example/demo/controller/expense/ExpenseController.java
package com.example.demo.controller.expense;

import java.util.List;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.example.demo.dto.expense.ExpenseCreateRequest;
import com.example.demo.entity.expense.Expense;
import com.example.demo.service.expense.ExpenseService;

import jakarta.validation.Valid;

@RestController
@RequestMapping("/api/expenses")
public class ExpenseController {

    private final ExpenseService expenseService;

    public ExpenseController(ExpenseService expenseService) {
        this.expenseService = expenseService;
    }

    @GetMapping
    public List<Expense> getAll() {
        return expenseService.getAll();
    }

    @GetMapping("/{id}")
    public ResponseEntity<Expense> getById(@PathVariable Long id) {
        return expenseService.getById(id)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<Expense> create(@Valid @RequestBody ExpenseCreateRequest request) {
        Expense created = expenseService.create(request);
        return ResponseEntity.status(HttpStatus.CREATED).body(created);
    }

    @PutMapping("/{id}")
    public ResponseEntity<Expense> update(@PathVariable Long id, @Valid @RequestBody ExpenseCreateRequest request) {
        return expenseService.update(id, request)
                .map(ResponseEntity::ok)
                .orElseGet(() -> ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        if (!expenseService.delete(id)) {
            return ResponseEntity.notFound().build();
        }
        return ResponseEntity.noContent().build();
    }
}

File: src/controllers/expense/ExpenseController.ts
import { ExpenseDTO, IExpenseService } from './Expense';

/**
 * Controller for Expense CRUD operations
 */
export class ExpenseController {
  private expenseService: IExpenseService;

  constructor(expenseService: IExpenseService) {
    this.expenseService = expenseService;
  }

  /**
   * Create a new expense
   */
  public async create(dto: ExpenseDTO): Promise<ExpenseCreateUpdateResponse> {
    const created = await this.expenseService.create(dto);
    return {
      id: created.id,
      title: created.title,
      amount: created.amount,
      date: created.date.toISOString(),
    };
  }

  /**
   * Get an expense by ID
   */
  public async getById(id: number): Promise<ExpenseCreateUpdateResponse | null> {
    const expense = await this.expenseService.getById(id);
    if (!expense) return null;
    return {
      id: expense.id,
      title: expense.title,
      amount: expense.amount,
      date: expense.date.toISOString(),
    };
  }

  /**
   * Update an existing expense
   */
  public async update(id: number, dto: ExpenseDTO): Promise<ExpenseCreateUpdateResponse | null> {
    const updated = await this.expenseService.update(id, dto);
    if (!updated) return null;
    return {
      id: updated.id,
      title: updated.title,
      amount: updated.amount,
      date: updated.date.toISOString(),
    };
  }

  /**
   * Delete an expense by ID
   */
  public async delete(id: number): Promise<boolean> {
    return await this.expenseService.delete(id);
  }

  /**
   * Get all expenses
   */
  public async getAll(): Promise<ExpenseCreateUpdateResponse[]> {
    const expenses = await this.expenseService.getAll();
    return expenses.map(expense => ({
      id: expense.id,
      title: expense.title,
      amount: expense.amount,
      date: expense.date.toISOString(),
    }));
  }
}